File: submission_002-mastermind/mastermind.py
import random


def run_game():
#step 1
    c_list = ["0", "0", "0", "0"]

    for i in range(0, len(c_list)):
        random_c_list = random.randint(1, 8)
        while random_c_list in c_list:
            random_c_list = random.randint(1, 8)
        c_list[i] = random_c_list
    
    print("4-digit Code has been set. Digits in range 1 to 8. You have 12 turns to break it.")
    
    
#step 2
    number_tries = 12
    
    while number_tries != 0:
        user_input = input("Input 4 digit code: ")
        
        while (user_input.isdigit() == False
               or len(user_input) < 4 
               or len(user_input) > 4
               or "0" in user_input
               or "9" in user_input):
            
            print("Please enter exactly 4 digits.")
            user_input = input("Input 4 digit code: ")
        
        user_input_list = list(user_input)

#step 3

        correct_i = 0
        incorrect_i = 0
        for i in range(0, len(c_list)):
            if c_list[i] == int(user_input_list[i]):
                correct_i += 1
            elif int(user_input_list[i]) in c_list:
                incorrect_i += 1
            else:
                i += 1
        print("Number of correct digits in correct place:    ", correct_i)
        print("Number of correct digits not in correct place:", incorrect_i)
        
        if correct_i == 4:
            print("Congratulations! You are a codebreaker!")
            print("The code was: ", end="")
            for i in range(0, len(c_list)):
                print(c_list[i], end="")
            print()
            break
        elif correct_i != 4:
            number_tries -= 1
            print("Turns left:", number_tries)
            
        if number_tries == 0:
            print("The code was: ", end="")    
            for i in range(0, len(c_list)):
                print(c_list[i], end="")
            print()
            break
    

if __name__ == "__main__":
    run_game()

File: mastermind/Mastermind.java
package za.co.wethinkcode.mastermind;

public class Mastermind {
    private final String code;
    private final Player player;

    public Mastermind(CodeGenerator generator, Player player){
        this.code = generator.generateCode();
        this.player = player;
    }
    public Mastermind(){
        this(new CodeGenerator(), new Player());
    }

    public void runGame(){
        //TODO: implement the main run loop logic

        System.out.println("4-digit Code has been set. Digits in range 1 to 8. You have 12 turns to break it.");

        String guess = player.getGuess();


        int tries = 12;

//        while(!(guess.equals("exit") || guess.equals("quit"))){
//            if (this.code.equals(guess)){
//                correctDigits += 1;
//            }else if(!(this.code.equals(guess))){
//                incorrectDigits += 1;
//            }
        while(!(guess.equals("exit") || guess.equals("quit"))){
            int correctDigits = 0;
            int incorrectDigits = 0;
            for(int i = 0; i < 4; i++){
                char a = this.code.charAt(i);

                if(guess.charAt(i) == this.code.charAt(i)){
                    correctDigits++;
                }else if(guess.contains(String.valueOf(a))){
                    incorrectDigits++;
                }
            }
        System.out.println("Number of correct digits in correct place: " + correctDigits);
        System.out.println("Number of correct digits not in correct place: " + incorrectDigits);

        if (tries == 1){
            System.out.println("No more turns left.");
            System.out.println("The code was: " + this.code);
            break;
        }else if(this.code.equals(guess)){
            System.out.println("Congratulations! You are a codebreaker!");
            System.out.println("The code was: " + this.code);
            break;
        }else if (!(guess.equals(this.code))){
            tries -= 1;
            System.out.println("Turns left: " + tries);
        }
//        System.out.println("Turns left: " + tries);
        guess = player.getGuess();
        }
    }

    public static void main(String[] args){
        Mastermind game = new Mastermind();
        game.runGame();
//        CodeGenerator code = new CodeGenerator();
//        code.generateCode();
//        Player player = new Player();
//        player.getGuess();
    }
}
